### PR TITLE
[terra-functional-testing] Added FileOutputReporter reporter

### LINKED ISSUE #634: [terra-functional-testing] Update the report writer to write the contents to a file
# Story

## Description

As a consumer of terra, I want to have my output written to a file (similar to what was done with the reporter in our old toolkit https://github.com/cerner/terra-toolkit-boneyard/blob/main/reporters/wdio/TerraWDIOSpecReporter.js)

## Motivation

So that this report can be utilized by a release site similarly to how the old toolkit's report was utilized

## Definition of Done

We'll know we're done when you can run wdio tests and the report can be pulled into our internal release site exactly the same way that things were done when toolkit was used.

### PR DESCRIPTION
**Pull request by sirishaNekkanti**
### Summary
Wdio File Output Reporter is a reporter that logs wdio test output to separate files based on locale, theme, and form-factor
Closes: #634

### CODE REVIEW

File: packages/terra-functional-testing/CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@
 * Added
   * Added `ignoreScreenshotMismatch` flag to disable test failure on snapshot mismatch.
   * Added configuration to retrieve and unzip screenshots from registry.
+  * Added FileOutputReporter reporter that logs wdio test output to separate files based on locale, theme, and form-factor
```
Comment (benbcai) [inline packages/terra-functional-testing/CHANGELOG.md]: This got added to the released section.

File: packages/terra-functional-testing/package.json [hunk 2]
```diff
@@ -73,7 +73,8 @@
     "uuid": "^3.0.0",
     "webdriverio": "^6.5.0",
     "webpack-dev-server": "^3.11.0",
-    "which": "^2.0.2"
+    "which": "^2.0.2",
+    "strip-ansi": "^6.0.0"
```
Comment (benbcai) [inline packages/terra-functional-testing/package.json]: This should be listed alphabetically.

File: packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# Wdio File Output Reporter
```
Comment (benbcai) [inline packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx]: Should this go inside the [Wdio Report](https://github.com/cerner/terra-toolkit/tree/fileoutput-reporter/packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/wdio-reporters.3) directory to group with [other reporters](https://engineering.cerner.com/terra-ui/dev_tools/cerner/terra-functional-testing/wdio-reporters/spec-reporter)?

File: packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+Wdio File Output Reporter is a reporter that logs wdio test output to separate files based on locale, theme, and form-factor with the following attributes
+
+- date and time of when the tests are started and when they are finished
```
Comment (benbcai) [inline packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx]:
```suggestion
- Date and time of when the tests are started and when they are finished
```

File: packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+- Locale, Form Factor, and Theme of the tests
+
+- Name of the tests and whether they've succeeded or failed
+
+**WDIO test output Directory**
+Wdio File Output Reporter logs the WDIO test output in tests/wdio/reports/ or test/wdio/reports/ depending on whether tests or test is the directory that contains specs
```
Comment (benbcai) [inline packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx]: Are we just going to support the `tests` directory or are we still continuing to support `test`?. I thought in one of the other PRs that we removed supporting `test`.

File: packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
+const fs = require('fs-extra');
+const path = require('path');
+const stripAnsi = require('strip-ansi');
+const SpecReporter = require('@wdio/spec-reporter').default;
+const endOfLine = require('os').EOL;
+const { Logger } = require('@cerner/terra-cli');
+
+const LOG_CONTEXT = '[Terra-Toolkit:fileOutput-reporter]';
```
Comment (benbcai) [inline packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js]: Most of our folders are named in snake case. We should avoid using both snake case with camel case.

File: packages/terra-functional-testing/CHANGELOG.md (hunk 1 above)
Comment (sirishaNekkanti) [reply packages/terra-functional-testing/CHANGELOG.md]: https://github.com/cerner/terra-toolkit/pull/665/files#diff-5121be21da689ee0a1bd5a529bbd578dc2bde92ce93b497f118e58290be63dc1R8-R10

File: packages/terra-functional-testing/package.json (hunk 2 above)
Comment (sirishaNekkanti) [reply packages/terra-functional-testing/package.json]: https://github.com/cerner/terra-toolkit/pull/665/files#diff-72c2847b50d4152bd910ef25c33dafbd31a04d507329a333036e6226e5e8ef71R73

File: packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx (hunk 3 above)
Comment (sirishaNekkanti) [reply packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx]: addressed https://github.com/cerner/terra-toolkit/pull/665/files#diff-11886e1f7aa3ee1e87b367b174084d9692573d2bc64322691e8e0b6d0ef1a898

File: packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx (hunk 4 above)
Comment (sirishaNekkanti) [reply packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx]: https://github.com/cerner/terra-toolkit/pull/665/files#diff-11886e1f7aa3ee1e87b367b174084d9692573d2bc64322691e8e0b6d0ef1a898R5

File: packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx (hunk 5 above)
Comment (sirishaNekkanti) [reply packages/terra-toolkit-docs/src/terra-dev-site/tool/terra-functional-testing/fileOutputReporter/FileOutputReporter.mdx]: https://github.com/cerner/terra-toolkit/pull/665/files#diff-11886e1f7aa3ee1e87b367b174084d9692573d2bc64322691e8e0b6d0ef1a898R12

File: packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js (hunk 6 above)
Comment (sirishaNekkanti) [reply packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js]: https://github.com/cerner/terra-toolkit/pull/665/files#diff-d8f1b66e8db21668273f6ad5ba417c4a5af04b1fcd0e46b5e2923ac1345fa743R1

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 7]
```diff
@@ -0,0 +1,207 @@
+const fs = require('fs-extra');
+const path = require('path');
+const stripAnsi = require('strip-ansi');
+const SpecReporter = require('@wdio/spec-reporter').default;
+const endOfLine = require('os').EOL;
+const { Logger } = require('@cerner/terra-cli');
+
+const LOG_CONTEXT = '[Terra-Toolkit:fileOutput-reporter]';
```
Comment (benbcai) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]:
```suggestion
const LOG_CONTEXT = '[terra-functional-testing:file-output-reporter]';
```

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    this.hasResultsDir();
+  }
+
+  /**
+   * Sets results directory for the test run. Uses the wdio reporterOptions.outputDir if set, otherwise
+   * it outputs to tests/wdio/reports.
+   * @return null;
+   */
+  setResultsDir() {
+    this.resultsDir = path.join(process.cwd(), 'tests', 'wdio', 'reports');
```
Comment (benbcai) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js:39]: Do we need to handle the reports path for mono-repos or will the report for all packages be in the root level?

Comment (sirishaNekkanti) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js:39]: reports for all the packages are in the root level

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    this.fileName = fileNameConf.join('-');
+  }
+
+  /**
+   * Set the package name to moduleName property if specsValue contains /package string
+   * @param {string} specsValue - File path of current spec file from runners
+   * @return null
+   */
+  setTestModule(specsValue) {
+    const index = specsValue.lastIndexOf(`packages${path.sep}`);
```
Comment (benbcai) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: Should we also add `${path.sep}` in front of `packages` to be more robust? Similar to [this](https://github.com/cerner/terra-toolkit/blob/8dedba1178b2993bd751dc0c35704f36c0bbbd0c/packages/terra-functional-testing/src/reporters/spec-reporter/wdio-spec-reporter.js#L120)?

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    if (index > -1) {
+      const testFilePath = specsValue.substring(index).split(path.sep);
+      const moduleName = testFilePath && testFilePath[1]
+        ? testFilePath[1]
+        : process.cwd().split(path.sep).pop();
```
Comment (benbcai) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: There was issue #597 in the spec reporter where the package name was incorrect when running on Jenkins CI. A change had to be made to parse the package name correctly. Does this potentially have the same issue that needs to be taken care of? This was the [fix](https://github.com/cerner/terra-toolkit/pull/613/files#diff-c4330edcf41e36606925cf520272bc65282b2454fee1d8d50c8c131ecc042ee6R123-R127).

cc @BenBoersma

Comment (benbcai) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: If `setTestModule` is doing the same thing as [getPackageName](https://github.com/cerner/terra-toolkit/blob/main/packages/terra-functional-testing/src/reporters/spec-reporter/wdio-spec-reporter.js#L118-L129), would it be better to make `getPackageName` reusable for both reporters?

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+          config: { ...runner.config, ...{ capabilities } },
+          sessionId: capabilities.sessionId,
+          isMultiremote: runner.isMultiremote,
+          instanceName,
+        }),
+      )
+      : this.getTestLink(runner);
+    const output = [
+      ...this.getHeaderDisplay(runner),
+      '',
```
Comment (benbcai) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js:164]: What is this empty string used for?

Comment (sirishaNekkanti) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js:164]: it is just used to format the data , I grabbed this functionality from the SpecReporter

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js (hunk 10 above)
Comment (sirishaNekkanti) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js:102]: I just realized setTestModule and getPackageName are doing two different things. setTestModule is grabbing the package name what it is running. where as getPackageName is getting the repo name

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+    if (runners && runners.length) {
+      runners.forEach((runner, index) => {
+        // determine correct file name given configuration for run
+        if (index === 0) {
+          const { cid, capabilities, config } = runner;
+          const { browserName } = capabilities;
+          this.fileNameCheck(config.launcherOptions, browserName, cid);
+        }
+
+        this.moduleName = this.setTestModule(runner.specs[0]);
```
Comment (benbcai) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: We probably don't need [line 24](https://github.com/cerner/terra-toolkit/pull/665/files#diff-d8f1b66e8db21668273f6ad5ba417c4a5af04b1fcd0e46b5e2923ac1345fa743R24) and this.moduleName can just be a regular const instead of a class var.

Comment (sirishaNekkanti) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: https://github.com/cerner/terra-toolkit/blob/8c50b42e3f8fdde85c9640506f1c51794677cb8d/packages/terra-functional-testing/src/reporters/file-output-reporter/index.js#L155

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        }
+      });
+    }
+  }
+
+  /**
+   * Formatting the filename based on locale, theme, and formFactor
+   * @return null
+   */
+  fileNameCheck({ formFactor, locale, theme }, browserName, cid) {
```
Comment (yuderekyu) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: I think the API would be much clearer if the parameters were condensed into single `options` or `configuration` object parameter, rather than a mixture of an object + two other parameters.

```suggestion
  fileNameCheck({ formFactor, locale, theme browserName, cid }) {
```

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+    const fileNameConf = ['fileOutput'];
+    if (locale) {
+      fileNameConf.push(locale);
+      this.resultJsonObject.locale = locale;
```
Comment (yuderekyu) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: it's not clear to me, via the method name, that this method is also creating the `resultsJsonObject`. my thought is to separate building the `this.resultsJsonObject` and building the `this.fileName` string into different methods, so we can separate concerns and keep logic simple.

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js (hunk 13 above)
Comment (yuderekyu) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: and I vote to rename this method to `buildFileName`

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+          config: { ...runner.config, ...{ capabilities } },
+          sessionId: capabilities.sessionId,
+          isMultiremote: runner.isMultiremote,
+          instanceName,
+        }),
+      )
+      : this.getTestLink(runner);
+    const output = [
+      ...this.getHeaderDisplay(runner),
+      '',
```
Comment (yuderekyu) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: my vote is to convert this multi line ternary into a regular if else statement, it's kind of hard to follow :/

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+    this.fileName = fileNameConf.join('-');
+  }
+
+  /**
+   * Set the package name to moduleName property if specsValue contains /package string
+   * @param {string} specsValue - File path of current spec file from runners
+   * @return null
+   */
+  // eslint-disable-next-line class-methods-use-this
```
Comment (benbcai) [inline packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: We should make `setTestModule` a static function and not have to disable this rule.

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js (hunk 13 above)
Comment (benbcai) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: Refactored the passed arguments: https://github.com/cerner/terra-toolkit/pull/665/commits/979bbe09908240c56f67a200b4457942fa47dda6

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js (hunk 14 above)
Comment (benbcai) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: Updated to set resultsJsonObject in a new function: https://github.com/cerner/terra-toolkit/commit/979bbe09908240c56f67a200b4457942fa47dda6

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js (hunk 15 above)
Comment (benbcai) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: Updated: https://github.com/cerner/terra-toolkit/commit/979bbe09908240c56f67a200b4457942fa47dda6

File: packages/terra-functional-testing/src/reporters/file-output-reporter/index.js (hunk 16 above)
Comment (benbcai) [reply packages/terra-functional-testing/src/reporters/file-output-reporter/index.js]: Made getPackageName static: https://github.com/cerner/terra-toolkit/commit/979bbe09908240c56f67a200b4457942fa47dda6

**Review by benbcai — APPROVED** [review]

**Review by mjhenkes — APPROVED** [review]

**Review by yuderekyu — APPROVED** [review]

**Comment by benbcai** [comment]
Thanks @yuderekyu for the +1!

**Review by pranav300 — APPROVED** [review]

**Review by ryanthemanuel — APPROVED** [review]
